### DISCUSSION TITLE
Chaining Streams

### DISCUSSION THREAD

**Discussion by weow147** [pschatzmann/arduino-audio-tools#1258] [Q&A] [answered]
I'd like to chain together multiple sources and sinks to create a vocal effects box.

1. i2s to fft (detect pitch)
2. fft to AudioEffectStream (use pitchshift to correct pitch if singer is out of tune, add user-selectable additional effects echo/reverb/etc)
3. AudioEffectStream to 3bandEQ (will attempt to modify 3band to 5band, or start from scratch and use filters)
4. 3bandEQ to is2 (done)

I've read the [detail instructions](https://github.com/pschatzmann/arduino-audio-tools/wiki/Introduction) and it seems i need to use [transformation/conversion classes](https://pschatzmann.github.io/arduino-audio-tools/group__transform.html) but i'm struggling.

**Comment by pschatzmann** [comment, marked as answer]
I think you did not get the chain right. 
The following should work:
```
                                      -> FFT (e.g. RealFFT)
I2SStream -copy-> MultiOutputStream
                                      -> AudioEffectStream -> Equilizer3Band -> I2SStream
```

**Reply by weow147** [reply]
I don't understand...
I2SStream i2s;  relies on a copier.  Input is hardware codec, output is destination of copier.  
AudioRealFFT fftc; also relies on a copier.  Input and output are specified by copier.
MultiOutput does not help.  it creates multiple parallel outputs, i need a sequential chain.


```
I2SStream i2s;  //relies on copier
AudioRealFFT fftc;  //relies on copier  
StreamCopy copier(fftc, i2s);  //i2s from codec copied to fft stream
AudioEffectStream audioStream(fftc);  //fft stream flows into audiostream
Equilizer3Bands eq3bands(audioStream);  //audiostream flows into eq
//how do i get back to i2s?  can i have TWO copiers?
```

I dont see how to put these LEGO together...

**Comment by pschatzmann** [comment]
No: AudioRealFFT is like any audio sink but can never be an audio source! To feed it you just define it as output of the MultiOutputStream along with the AudioEffectStream. Here is the code for the chain described above:

```
I2SStream i2s;  // source and destination
AudioRealFFT fft;  // use custom logic to adjust effects  
MultiOutputStream multi; // output to effect and fft
Equilizer3Bands eq3bands(i2s);  // equilizer to i2s
AudioEffectStream effect(eq3bands);  // effect to eq
StreamCopy copier(multi, i2s);  // copy data to effect and fft

void setup() {
  multi.add(fft);
  multi.add(effect);
...

}
```

**Reply by weow147** [reply]
i spent about 20 minutes thinking this over, it makes sense now, thankyou.
i couldn't get my head around fft as a dead-end.  but it IS a dead-end, correctly.  it also provides a data point that i will use with pitchshift later.  The rest is a simple cascade.

I cant wait to work on this over the weekend!  i'll let you know when i get it working...
